sympy/matrices/common.py

```diff
@@ -2621,7 +2653,11 @@ def _matrixify(mat):
         return mat
     if hasattr(mat, 'shape'):
         if len(mat.shape) == 2:
-            return _MatrixWrapper(mat)
+            try: # test indexing behavior of matrix-like object
+                e = mat[0][0]
+                return _MatrixWrapperRowIndexing(mat)
+            except (TypeError, IndexError, KeyError, ValueError):
+                return _MatrixWrapper(mat)
```

Comment:
Why is this catching so many exceptions? Which are actually needed?

What classes of matrices are you trying to distinguish?

Reply:
I think that this class was intended to work support any arbitrary matrices, so errors can be ambiguous.
Please add some optional dependency tests that proves that this works with some numpy, scipy, mpmath matrices.

Reply:
Not making any assumptions about the class and just trying to cover all the reasons indexing can fail in this part:
```
>>> 1 [0] [0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not subscriptable

>>> np.ndarray ((0, 0)) [0] [0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: index 0 is out of bounds for axis 0 with size 0

>>> {(0,0): 1} [0] [0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 0
```
`ValueError` is included just in case, the idea being that an indexing error in a matrix type test shouldn't propagate up.

The original issue is "_MatrixWrapper does not wrap iteration of numpy matrix properly" and this is the fix for that particular problem. But if you ask me though this whole wrapper class is a poor quickly implemented band-aid. I don't see how it can work for anything other than carrying out SymPy operations on numpy or other matrix source data. The `__getattr__` returning the source matrix attributes will return their own representations of the result which will not be compatible with internal SymPy functions since those assume simpified returns. The `__getitem__` should obviously sympify as well as well as a `__setitem__` which coerces to source types.

A true wrapper class would translate parameters going into source matrix functions as well as the returns, as well as adding known foreign matrix types to the `sympify` function so they can be coerced and used across all SymPy operations where SymPy matrices are expected.

Reply:
I think it would be better to special case the known expected types here rather than catch the exception.

In any case can we make sure that the smallest possible amount of code is wrapped by the `try`? I don't think that the `return _MatrixWrapperRowIndexing(mat)` needs to be there.